sources/osg/CullVisitor.js

```diff
@@ -375,49 +378,21 @@ MACROUTILS.createPrototypeObject(
                 this.applyFunctionArray[node.nodeTypeID].call(this, node);
             },
 
-            createOrReuseRenderStage: function(classInstance) {
-                var type = !classInstance ? 'RenderStage' : classInstance.className();
-                var classCtor = !classInstance ? RenderStage : classInstance.constructor;
-
-                var stack;
-                if (this._reserveRenderStageStacks[type]) {
-                    stack = this._reserveRenderStageStacks[type];
-                } else {
-                    stack = new osgPool.OsgObjectMemoryStack(classCtor);
-                    this._reserveRenderStageStacks[type] = stack;
-                }
-                return stack.get().init();
-            },
-
             createOrReuseRenderLeaf: function() {
-                var l = this._reserveLeafStack[this._reserveLeafStackCurrent++];
-                if (this._reserveLeafStackCurrent === this._reserveLeafStack.length) {
-                    this._reserveLeafStack.push(new RenderLeaf());
-                }
-                return l;
-            },
-
-            resetRenderLeafStack: function() {
-                for (var i = 0, j = this._reserveLeafStackCurrent; i <= j; i++) {
-                    this._reserveLeafStack[i].reset();
-                }
+                return this._pooledLeaf.getOrCreateObject();
             },
 
-            createOrReuseCullSettings: function() {
-                var l = this._reserveCullSettingsStack[this._reserveCullSettingsStackCurrent++];
-
-                if (
-                    this._reserveCullSettingsStackCurrent === this._reserveCullSettingsStack.length
-                ) {
-                    this._reserveCullSettingsStack.push(new CullSettings());
-                }
-                return l;
-            },
+            createOrReuseRenderStage: function(classInstance) {
+                var type = !classInstance ? 'RenderStage' : classInstance.className();
 
-            resetCullSettingsStack: function() {
-                for (var i = 0, j = this._reserveCullSettingsStackCurrent; i <= j; i++) {
-                    this._reserveCullSettingsStack[i].reset();
+                if (!this._pooledRenderStages[type]) {
+                    var classCtor = !classInstance ? RenderStage : classInstance.constructor;
+                    var createRenderStage = function() {
+                        return new classCtor();
+                    };
+                    this._pooledRenderStages[type] = new PooledResource(createRenderStage);
                 }
+                return this._pooledRenderStages[type].getOrCreateObject().init();
```

Comment:
I think it could be better if the PooledResource always return clean Objects when getOrCreate(). 
Is there any case when we don't want a clean/reseted object? 
Simply forcing to every poolable object to have a reset method does the trick to call reset in getOrCreateObject(), and removes the hack of doing twice init() things.

Otherwise we should state very clear in the documentation that returned objects are not clean, and user is responsible to reset/init them

Reply:
agree but for performance reason it's the user who needs to clean the object he gets
I will add a comment in the source code